### PR TITLE
feat(reid): appearance ReID association via standalone reid package

### PR DESCRIPTION
**Pull request by AlexBodner**
## Summary

Adds optional appearance-ReID to BoT-SORT, sourcing the model stack from the standalone [`reid`](https://github.com/roboflow/re-ID) package. Trackers ships only NumPy association glue (`ReIDEncoder`, `FeatureBank`, `appearance_similarity`, `extract_detection_embeddings`) plus BoT-SORT appearance-IoU fusion. Encoder weights, preprocessing, catalog, and gallery evaluation live in `reid` and are never vendored.

This branch is merged with current `develop` (including McByte) and is independent of the earlier stacked ReID PRs (#480/#501/#502).

## Added things

- **Deps:** `trackers[reid]` optional extra → `reid>=0.1.0.dev0,<0.2` (PyPI). `reid.*` in mypy overrides; `uv.lock` updated.
- **Association glue** (`src/trackers/core/reid/`, torch-free): `ReIDEncoder`, `FeatureBank` (BoT-SORT EMA L2 before/after blend), `appearance_similarity`, `extract_detection_embeddings`.
- **Lazy import:** CLI loads `reid.ReIDModel` at use time (same pattern as detection); importing BoT-SORT stays free of `reid` / torch (isolation test).
- **BoT-SORT wiring:** appearance-IoU fusion, per-track feature bank, params `reid_model`, `reid_ema_alpha`, `reid_appearance_threshold`, `reid_proximity_threshold`.
- **CLI:** `--tracker.reid.{enable,model,device,architecture}` on `trackers track` (BoT-SORT only).
- **Tuning:** `trackers tune` and `Tuner` search `reid_appearance_threshold` and `reid_proximity_threshold` when an encoder is passed (`--reid.model`, or `reid_model` in `fixed_params`). Search-space entries gain an optional `requires` key, a `search_space` override is accepted per run, and each detection is embedded once per study instead of once per trial.
- **Tests:** association / fusion / CLI unit tests (no `--extra reid` required), `@integration` smoke with a real `reid.ReIDModel`, isolation test.
- **Docs / notebook:** `docs/learn/reid.md` (threshold guide + MOT17/SoccerNet association-local plots), `docs/api/reid.md`, install notes, BoT-SORT Appearance ReID section; `notebooks/eval_trackers_reid.ipynb` installs from PyPI.
- **CI:** unit (`ci-tests.yml`) syncs without `--extra reid`; integration (`ci-integrations.yml`) keeps `--extra reid`.

## Benchmarks

BoT-SORT with and without ReID, fused as in the BoT-SORT paper (`min(d_iou, d_app)`). Within each dataset both rows share detections, CMC and the docs' tuned motion parameters; only the appearance branch differs. For MOT17, DanceTrack and SoccerNet, both ReID thresholds were tuned together on the tuning split with the `Tuner` added here (20 trials; SoccerNet stopped at 13 by a 7 h cap). The SportsMOT threshold comes from a sweep at the default proximity threshold, and the fine-tuned SoccerNet thresholds from a grid on held-out sequences. Each pick was scored once on test.

**Generic encoders** (not trained on the evaluated dataset)

| Dataset | Encoder | ReID parameters | Tuning split HOTA | Test HOTA | Test IDF1 | Test MOTA |
| --- | --- | --- | --: | --: | --: | --: |
| MOT17 | `fastreid_mot17_sbs50` | `reid_appearance_threshold=0.10`, `reid_proximity_threshold=0.5` | 69.05 → 69.64 (val-half) | 63.86 → 63.85 | 78.74 → 78.82 | 79.42 → 79.41 |
| SportsMOT | `fastreid_mot17_sbs50` | `0.30`, `0.5` | 82.00 → 82.69 (val) | 74.15 → **75.62** | 74.06 → **75.66** | 96.89 → 96.90 |
| DanceTrack | `fastreid_mot17_sbs50` | `0.48`, `0.51` | 53.89 → 58.34 (val) | 57.8 → 57.6 | 57.9 → 57.6 | 92.2 → 92.1 |
| SoccerNet | `fastreid_mot17_sbs50` | `0.05`, `0.35` | 86.95 → 86.98 (train) | 85.00 → 84.96 | 79.68 → 79.66 | 97.25 → 97.25 |

**Fine-tuned encoders** (`osnet_x1_0` trained on the dataset's train split)

| Dataset | ReID parameters | Tuning split HOTA | Test HOTA | Test IDF1 | Test MOTA |
| --- | --- | --: | --: | --: | --: |
| MOT17 | `0.25`, `0.5` | 69.05 → 69.00 (val-half) | 63.8 → **64.12** | 78.7 → **79.16** | 79.4 → 79.36 |
| SoccerNet | `0.075`, `1.0` | 85.72 → 88.97 (12 held-out train sequences) | 85.00 → **87.30** | 79.68 → **83.16** | 97.25 → **98.73** |

Every generic encoder gains on its tuning split; only SportsMOT keeps the gain on test, and the rest land within 0.2 HOTA of BoT-SORT without ReID. Fine-tuning is what makes SoccerNet the largest gain. The SoccerNet encoder was trained on the first 45 of the 57 train sequences, so its thresholds were selected on the last 12; that search picks the same values the earlier study found. The MOT17 fine-tuned row keeps the default thresholds, and its encoder was trained on the first half of the same sequences that val-half evaluates.

With the BoT-SORT paper's own MOT17 settings (the library defaults) instead of the tuned ones, the same FastReID encoder at `0.25`, `0.5` scores 63.94 on MOT17 test against 63.42 without ReID (+0.52; val-half +0.30). The paper's per-sequence overrides for test sequences are not applied.

Test scoring: Codabench for MOT17, SportsMOT and DanceTrack; SoccerNet locally with `trackers eval` against the public test ground truth, all 49 sequences checked complete. Where each number was produced (device, code version) is recorded per run; the same configuration moves by up to 0.1 HOTA between an Apple GPU and CUDA, and reproduces exactly on the same device class. `reid_model=None` is the default, so existing benchmark numbers are unchanged.

## Pre-merge checklist

- [x] Publish `reid` to PyPI (`0.1.0.dev0`).
- [x] Swap the git pin in `pyproject.toml` to a PyPI range (`reid>=0.1.0.dev0,<0.2`) and refresh `uv.lock`.
- [x] Resolve merge conflicts with `develop` (McByte / mask extras) so PR CI can run.
- [x] Confirm integration CI is green with `--extra reid` (smoke should run, not skip).
- [x] Refresh MOT17 / SoccerNet appearance-threshold docs figures (association-local sampling).
- [x] pre-commit.ci green.

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -47,6 +47,8 @@ dependencies = [
 [project.optional-dependencies]
 detection = ["inference-models>=0.19.0"]
 tune = ["optuna>=3.0.0"]
+# Temporary git pin until reid is on PyPI. Swap to reid>=0.1.0,<0.2 before merge.
+reid = ["reid @ git+https://github.com/roboflow/re-ID.git@main"]
```
Comment (Borda) [inline pyproject.toml]:
```suggestion
reid = ["reid"]
```

**Comment by AlexBodner** [comment]
@Borda on the \`reid = [\"reid\"]\` suggestion: keeping the floor/ceiling pin (\`reid>=0.1.0.dev0,<0.2\`) for now so installs stay on the published pre-release and do not silently float to an incompatible 0.2+. Happy to widen once we cut a stable \`0.1.0\`.

File: docs/api/reid.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+# ReID API
+
+Requires the `reid` extra (`pip install "trackers[reid]"`, see the
+[install guide](../learn/install.md)).
+
+This page covers the `ReIDEncoder` protocol, `FeatureBank`, and appearance
+association helpers in `trackers.core.reid`. For enabling appearance on
+BoT-SORT, threshold selection, and benchmark results, see the
+[ReID appearance guide](../learn/reid.md). Model loading and gallery evaluation
+are in the standalone [`reid`](https://github.com/roboflow/re-ID) package.
```
[the remark this answers is not in the document]
Comment (AlexBodner) [reply docs/api/reid.md]: dont think its worth the distinction as for using reid functionality we need the extra install and model loading is the base for reid

File: docs/learn/reid.md [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+
+```python
+from reid import ReIDModel
+
+from trackers import BoTSORTTracker
+
+reid_model = ReIDModel.from_pretrained("fastreid_mot17_sbs50")
+tracker = BoTSORTTracker(reid_model=reid_model, appearance_threshold=0.2)
+```
+
```
[the remark this answers is not in the document]
Comment (AlexBodner) [reply docs/learn/reid.md:39]: applied in a nicer way

File: notebooks/eval_trackers_reid.ipynb [hunk 4]
```diff
@@ -0,0 +1,1218 @@
...
+                "try:\n",
+                "    import google.colab  # noqa: F401\n",
+                "\n",
+                "    IN_COLAB_INSTALL = True\n",
+                "except ImportError:\n",
+                "    IN_COLAB_INSTALL = False\n",
+                "\n",
+                "if IN_COLAB_INSTALL:\n",
+                "    # Feature branch until trackers[reid] is on PyPI develop/main.\n",
+                "    TRACKERS_GIT = \"trackers[reid] @ git+https://github.com/roboflow/trackers.git@feat/core/reid-consume-reid-package\"\n",
```
[the remark this answers is not in the document]
Comment (AlexBodner) [reply notebooks/eval_trackers_reid.ipynb]: fixed

File: src/trackers/core/reid/fusion.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+Fusion methods are numpy-only and take track-detection similarity matrices, so they are reusable across trackers rather
+than tied to any single one.
+"""
+
+from __future__ import annotations
+
+import numpy as np
+
+
+def fuse_botsort_reid_association(
```
Comment (tstanczyk95) [inline src/trackers/core/reid/fusion.py:24]: `fuse_botsort_reid_association()` is public-ish NumPy glue and does not itself validate that the incoming association similarity lies in [0,1]. It is done outside, but not here. Maybe it's worth to add such a validation here as well?

File: src/trackers/core/reid/fusion.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+            the fused association.
+        proximity_iou_similarity: Standard-IoU similarities with shape ``(T, N)``.
+            Defaults to ``association_similarity``.
+
+    Returns:
+        Fused track-detection similarities with shape ``(T, N)``, obtained from
+        ``1 - min(d_iou, d_app)`` after applying both gates.
+    """
+    if proximity_iou_similarity is None:
+        proximity_iou_similarity = association_similarity
```
Comment (tstanczyk95) [inline src/trackers/core/reid/fusion.py:59]: So can it happen that `ciou` or `diou` will be used as `proximity_iou_similarity`? 

Then it might cause some unintended behavior with values outside the ranges designed for original BoT-SORT.

File: src/trackers/core/botsort/tracker.py
Comment (tstanczyk95) [inline src/trackers/core/botsort/tracker.py:59]: With ReID enabled, `minimum_iou_threshold_first_assoc` is also applied to the fused association similarity. Could the docs clarify this? The current BoT-SORT parameter table describes it as IoU x detection confidence, and minimum_iou_threshold_unconfirmed_assoc as a minimum IoU, although appearance also participates in those stages.

File: docs/guides/reid.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+
+    from trackers import BoTSORTTracker
+
+    reid_model = ReIDModel.from_pretrained("fastreid_mot17_sbs50")
+    tracker = BoTSORTTracker(reid_model=reid_model, reid_appearance_threshold=0.2)
+    ```
+
+=== "CLI"
+
+    `--reid.model` takes a curated alias, an `hf://` URL, or a local path, and enables appearance association on its own. Tracker parameters keep their Python names, so the appearance gate is `--tracker.reid_appearance_threshold`. See the [CLI reference](cli.md) for every argument.
```
Comment (tstanczyk95) [inline docs/guides/reid.md:44]: Maybe we should be careful here?

As for CLI we have such naming for minimal values:
`minimum_iou_threshold` : `--tracker.min_iou_threshold`

See here : https://github.com/roboflow/trackers/blob/a01aef5010accf65c3a566397ec946aacb36b8d5/src/trackers/cli/track.py#L151

File: docs/trackers/botsort.md
Comment (tstanczyk95) [inline docs/trackers/botsort.md:65]: Maybe we can also mention this for the reID now?

**Review by tstanczyk95 — COMMENTED** [review]
@AlexBodner Thanks a lot, this is a very good and solid piece of work!

I have some remarks to further improve it and to clarify it. See them below as well as in the code.

**1.**
Do we have the capacity and data to get the results when the re-ID model is tuned on DanceTrack and (separately) on SportsMOT?

It would be very useful to see that. Also for the next development parts.
Besides could we see with the numbers if the `osnet_x1_0_msmt17_combineall` helps or not on SoccerNet and MOT17?

**2.**
Can we add a test with a BoT-SORT instantiated as`reid_model=None` to see if it gives exactly the pre-PR BoT-SORT behavior so as to assure the backwards compatibility?

**3.**
Can we also a test for fusion to avoid see if we face and/or handle such cases as `association_similarity=1.2`, as mentioned in the code review [here](https://github.com/roboflow/trackers/pull/511/changes#r3884426036) and [here](https://github.com/roboflow/trackers/pull/511/changes#r3884432159)?

**4. - Colab [notebook](https://github.com/roboflow/trackers/blob/a01aef5010accf65c3a566397ec946aacb36b8d5/docs/cookbooks/how-to-add-reid-to-trackers.ipynb) remarks**

**a.** Let’s put some reference links at the beginning of the notebook? So that the user knows where to look for more details instantly. E.g. if they want to learn more about that gating mentioned in `Choose an appearance threshold`

**b.** Maybe we can also mention about `reid_proximity_threshold` here? What it does, what is it for.

**c.** 
> This notebook enables ReID on BoT-SORT with the [reid](https://reid.roboflow.com/latest/) package

Not sure if the current link at the top will be working at that time. As it leads to reID pages, which are not available yet.

**d.** 
We could also gently add the links to the reID training procedure (so that one can use their own models) - once it becomes available.
Not to overwhelm the user, but to let them not that such a cool stuff is already there ;)

File: docs/guides/reid.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+
+Two things follow. First, a threshold validated on adjacent frames says little about re-association: at θ=0.2 appearance helps 98% of same-ID pairs one frame apart but only 51% across the default 30-frame lost-track buffer. Second, the price of a tight θ over long gaps is missed re-associations rather than extra wrong ones, because the different-ID rate stays near 1% throughout. If you raise `lost_track_buffer` to recover tracks after long occlusions, raise `reid_appearance_threshold` with it and re-check the different-ID column.
+
+The cross-domain encoder fails differently. On SoccerNet the different-ID rate at θ=0.2 stays between 44% and 51% at every gap, so the frame gap is not what limits it; the encoder simply cannot separate players in matching kits at any horizon. Widening the gap costs same-ID pairs (99.6% down to 87.0%) without ever making the different-ID side usable, which is why θ has to come down to about 0.1 on this domain instead of being traded against the gap.
+
+![OSNet MSMT17 separability vs frame gap](../assets/reid/soccernet-osnet-appearance-distances-vs-gap.png)
+
+---
+
+## BoT-SORT with and without ReID
```
Comment (tstanczyk95) [inline docs/guides/reid.md:149]: Could we also have it updated with the re-ID tuned on the datasets?
(See the remark **1.** from the first review comment [here](https://github.com/roboflow/trackers/pull/511#pullrequestreview-5055669987))

File: .github/workflows/ci-reid-prerelease.yml [hunk 9]
```diff
@@ -0,0 +1,31 @@
+name: ReID Pre-release Compatibility
```
Comment (Borda) [inline .github/workflows/ci-reid-prerelease.yml:1]:
```suggestion
name: ReID Compatibility
```
So we do not have multiple workflows later

File: .github/workflows/ci-reid-prerelease.yml [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+        uses: astral-sh/setup-uv@c771a70e6277c0a99b617c7a806ffedaca235ff9  # v9.0.0
+        with:
+          python-version: "3.12"
+          activate-environment: true
+
+      - name: 🚀 Install Packages with the latest allowed reid pre-release
+        run: uv sync --group dev --extra reid --upgrade-package reid
+
+      - name: 🧪 Smoke-test the trackers to reid boundary
+        run: uv run pytest tests/core/test_botsort_reid.py::TestBoTSORTTrackerReID::test_real_reid_model_runs_over_frames -v --tb=short
```
Comment (Borda) [inline .github/workflows/ci-reid-prerelease.yml:31]: Rather use pytest markers, since accidnet rename would brek this hardcoded link

File: docs/api/reid.md [hunk 11]
```diff
@@ -0,0 +1,75 @@
+---
+description: Python API reference for the ReID encoder protocol, feature bank, appearance association utilities, and threshold-selection plots in Roboflow Trackers.
```
Comment (Borda) [inline docs/api/reid.md:2]: Is it missing a name?

File: docs/api/reid.md [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+    sample_appearance_distances,
+    sweep_frame_gap,
+)
+
+embeddings, ids, frame_ids, sequence_ids = extract_ground_truth_embeddings(model, "mot17/val", keep_classes=(1,))
+distances = sample_appearance_distances(embeddings, ids, frame_ids, sequence_ids)
+same_id_rate, different_id_rate = distances.rates_at(0.25)
+plot_appearance_distances(distances, thresholds={0.20: "selected", 0.25: "default"})
+plot_frame_gap_sweep(sweep_frame_gap(embeddings, ids, frame_ids, sequence_ids))
+```
```
Comment (Borda) [inline docs/api/reid.md:47]: All others are just linking to the API, so this feels quite lost; however, it is important. Shall we rather have a separate page with how to calibrate or configure re-id?

File: docs/assets/reid/mot17-fastreid-appearance-distances-vs-gap.png
Comment (Borda) [inline docs/assets/reid/mot17-fastreid-appearance-distances-vs-gap.png:1]: Could we export these as SVG, so we won't compromise quality for amount of KB

File: docs/cookbooks/how-to-add-reid-to-trackers.ipynb
Comment (Borda) [inline docs/cookbooks/how-to-add-reid-to-trackers.ipynb:1]: Could we rather use rfdert instead of yolox?

File: docs/guides/reid.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+
+=== "Python"
+
+    ```python
+    from reid import ReIDModel
+
+    from trackers import BoTSORTTracker
+
+    reid_model = ReIDModel.from_pretrained("fastreid_mot17_sbs50")
+    tracker = BoTSORTTracker(reid_model=reid_model, reid_appearance_threshold=0.2)
```
Comment (Borda) [inline docs/guides/reid.md:39]:
```suggestion
    tracker = BoTSORTTracker(reid_model=reid_model, reid_config=...)
```
whihc would be dataclass os we wont explode the args for trackers

File: docs/guides/reid.md [hunk 14]
```diff
@@ -0,0 +1,179 @@
...
+
+| Frame gap  | ROC AUC | same-ID below 0.2 | different-ID below 0.2 |
+| :--------- | :-----: | :---------------: | :--------------------: |
+| 1          |  0.998  |       98.0%       |          1.7%          |
+| 2 to 5     |  0.987  |       87.6%       |          1.6%          |
+| 6 to 15    |  0.957  |       67.4%       |          1.1%          |
+| 16 to 30   |  0.929  |       51.4%       |          1.1%          |
+| 31 to 60   |  0.899  |       39.8%       |          0.9%          |
+| 61 to 120  |  0.865  |       31.8%       |          0.8%          |
+| 121 to 240 |  0.854  |       28.7%       |          0.8%          |
```
Comment (Borda) [inline docs/guides/reid.md:137]: tbh, not sure how to read this table or what it tells me...
(if the gap just one  big or ir random dropped frames?)

File: docs/guides/reid.md [hunk 15]
```diff
@@ -0,0 +1,179 @@
...
+
+### SoccerNet test (OSNet MSMT17)
+
+Oracle detections, CMC on, SoccerNet-tracking test (same protocol as the [benchmark results](../evaluations/results.md) default table). ReID: `osnet_x1_0_msmt17_combineall` (MSMT17 pretrained), so this is a cross-domain encoder on soccer footage.
+
+| Config                          |   HOTA   |   IDF1   |   MOTA   |
+| :------------------------------ | :------: | :------: | :------: |
+| BoT-SORT                        |   84.5   |   79.3   | **96.6** |
+| BoT-SORT + OSNet MSMT17 (θ=0.2) | **84.6** | **79.4** | **96.6** |
+| BoT-SORT + OSNet MSMT17 (θ=0.1) |   82.9   |   77.7   |   96.5   |
```
Comment (Borda) [inline docs/guides/reid.md:179]: Let's also add one col with wall time so it's clear what the cost is for use re-id

File: src/trackers/cli/track.py [hunk 16]
```diff
@@ -65,6 +65,32 @@ class DetectionOptions:
     api_key: str | None = None
 
 
+@dataclass
```
Comment (Borda) [inline src/trackers/cli/track.py:68]: Shall it be frozen?

File: src/trackers/core/botsort/tracker.py [hunk 17]
```diff
@@ -83,13 +87,26 @@ class BoTSORTTracker(BaseTracker):
...
+            matches the geometry-only BoT-SORT baseline.
+        reid_ema_alpha: EMA momentum for track appearance features. Default ``0.9``.
+        reid_appearance_threshold: Appearance distance gate. Drops the appearance term
+            when the halved cosine distance ``0.5 * (1 - cos_sim)`` exceeds this
+            value, leaving the pair scored on geometry alone. Default ``0.25``
+            (BoT-SORT ``appearance_thresh``).
+        reid_proximity_threshold: Standard-IoU distance gate applied before appearance
+            is used. Computed from true IoU even when ``iou`` is GIoU/DIoU/CIoU.
+            Default ``0.5`` (BoT-SORT ``proximity_thresh``; requires
+            ``IoU >= 1 - reid_proximity_threshold``).
```
Comment (Borda) [inline src/trackers/core/botsort/tracker.py:102]: I would rather wrap them into one config...

File: src/trackers/core/botsort/tracker.py [hunk 18]
```diff
@@ -181,11 +213,15 @@ def update(
...
         Notes:
-            - If CMC is enabled, pass the current video frame via ``frame`` so the
-              tracker can estimate a global affine transform and warp predicted
-              track states before association. When ``frame=None`` and
-              ``enable_cmc=True``, CMC is silently skipped for that step.
+            - If CMC or ReID is enabled, pass the current video frame via ``frame``
+              so the tracker can estimate a global affine transform and/or extract
+              appearance embeddings before association. When ``frame=None`` and
+              ``enable_cmc=True``, CMC is silently skipped for that step; ReID
+              requires ``frame`` and raises when it is ``None``.
```
Comment (Borda) [inline src/trackers/core/botsort/tracker.py:220]: let's set line-len as 120 for all to have consistent docs and code

File: src/trackers/core/botsort/tracker.py [hunk 19]
```diff
@@ -254,38 +290,47 @@ def update(
         # redundant, so all stages read boxes from this map keyed by ``id()``.
         predicted_state_boxes = {id(track): track.get_state_bbox() for track in self.tracks}
 
+        det_embeddings: np.ndarray | None = None
+        if self.reid_model is not None and frame is not None:
+            if len(high_boxes) > 0:
```
Comment (Borda) [inline src/trackers/core/botsort/tracker.py:295]: Why not one if together?

File: src/trackers/core/reid/appearance.py [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+        FileNotFoundError: If no sequence under ``dataset_root`` has a
+            ``gt/gt.txt``, or a named sequence is missing one.
+        ValueError: If no crop survived the filters.
+
+    Examples:
+        >>> from trackers.core.reid import extract_ground_truth_embeddings  # doctest: +SKIP
+        >>>
+        >>> crops = extract_ground_truth_embeddings(  # doctest: +SKIP
+        ...     model, "mot17/val", keep_classes=(1,)
+        ... )
```
Comment (Borda) [inline src/trackers/core/reid/appearance.py:133]: doctest which is copletely skipped doe not make any sense

File: src/trackers/core/reid/appearance.py [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+    embeddings: list[np.ndarray] = []
+    ids: list[int] = []
+    frame_ids: list[int] = []
+    sequence_ids: list[int] = []
+    identity_by_key: dict[str, int] = {}
+
+    for sequence_id, name in enumerate(names):
+        ground_truth = load_mot_file(root / name / "gt" / "gt.txt")
+        frame_dir = root / name / "img1"
+        for frame_id in range(1, max(ground_truth) + 1, frame_stride):
```
Comment (Borda) [inline src/trackers/core/reid/appearance.py:149]: This could be a generator if we plan to use it in multiple places

File: src/trackers/core/reid/appearance.py [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+    )
+
+
+def appearance_similarity(
+    track_features: Sequence[np.ndarray | None],
+    det_embeddings: np.ndarray,
+    *,
+    det_embeddings_normalized: bool = False,
+) -> np.ndarray:
+    """Compute cosine similarities between track and detection embeddings.
```
Comment (Borda) [inline src/trackers/core/reid/appearance.py:186]: What is the value range, and is higher better?

File: src/trackers/core/reid/thresholds.py [hunk 23]
```diff
@@ -0,0 +1,543 @@
...
+    """Yield ``(value, line keyword arguments)`` for each reference line."""
+    notes = thresholds if isinstance(thresholds, Mapping) else {}
+    for index, value in enumerate(thresholds):
+        note = notes.get(value)
+        color, linestyle, linewidth = _THRESHOLD_STYLES[min(index, len(_THRESHOLD_STYLES) - 1)]
+        label = f"θ = {value:.2f}" + (f" ({note})" if note else "")
+        yield value, {"label": label, "color": color, "ls": linestyle, "lw": linewidth}
+
+
+def plot_appearance_distances(
```
Comment (Borda) [inline src/trackers/core/reid/thresholds.py:399]: I would lean towad create a visual module with all these plotting

File: src/trackers/core/reid/fusion.py (hunk 6 above)
Comment (AlexBodner) [reply src/trackers/core/reid/fusion.py:59]: if the user wants it, but didnt benchmark it. We can remove it

File: src/trackers/core/reid/fusion.py (hunk 5 above)
Comment (AlexBodner) [reply src/trackers/core/reid/fusion.py:24]: it is a function intended for use only inside the trackers which already have all the harness for that. It would be redundant I think. Do you see it scaling more and being worth the checks?

File: docs/cookbooks/how-to-add-reid-to-trackers.ipynb
Comment (AlexBodner) [reply docs/cookbooks/how-to-add-reid-to-trackers.ipynb:1]: that were @tstanczyk95 detections, replicating bytetrack  paper approach. We could maybe do it now, not sure worth the effort

File: docs/guides/reid.md (hunk 13 above)
Comment (AlexBodner) [reply docs/guides/reid.md:39]: sounds good I can try that

File: docs/guides/reid.md (hunk 14 above)
Comment (AlexBodner) [reply docs/guides/reid.md:137]: will explain it in text next to it then :)

File: docs/guides/reid.md (hunk 15 above)
Comment (AlexBodner) [reply docs/guides/reid.md:179]: yes, sounds good. but which hardware? going for colab T4 unless you think my local mac m4 pro is better

File: src/trackers/core/reid/fusion.py (hunk 6 above)
Comment (AlexBodner) [reply src/trackers/core/reid/fusion.py:59]: so, to make it simpler I now disabled the possibility of using IoU variant with Reid enabled. Please confirm with me if its fine @tstanczyk95

File: src/trackers/core/reid/thresholds.py (hunk 23 above)
Comment (AlexBodner) [reply src/trackers/core/reid/thresholds.py:399]: @Borda how would you do that? another file just for the plots or other complete part of the library?

Comment (Borda) [reply src/trackers/core/reid/thresholds.py:399]: Aggregate all visual function into own module...

File: src/trackers/core/botsort/tracker.py (hunk 17 above)
Comment (AlexBodner) [reply src/trackers/core/botsort/tracker.py:102]: done

File: docs/guides/reid.md (hunk 13 above)
Comment (AlexBodner) [reply docs/guides/reid.md:39]: done

`-    from trackers import BoTSORTTracker
+    from trackers import BoTSORTTracker, ReIDConfig

     reid_model = ReIDModel.from_pretrained("fastreid_mot17_sbs50")
-    tracker = BoTSORTTracker(reid_model=reid_model, reid_appearance_threshold=0.2)
+    tracker = BoTSORTTracker(reid_model=reid_model, reid_config=ReIDConfig(appearance_threshold=0.2))`

File: src/trackers/core/reid/fusion.py (hunk 5 above)
Comment (AlexBodner) [reply src/trackers/core/reid/fusion.py:24]: now its explicit in the contract

File: src/trackers/core/botsort/tracker.py
Comment (AlexBodner) [reply src/trackers/core/botsort/tracker.py:59]: clarified on docs

File: docs/trackers/botsort.md
Comment (AlexBodner) [reply docs/trackers/botsort.md:65]: yes, done

File: src/trackers/core/botsort/tracker.py (hunk 19 above)
Comment (AlexBodner) [reply src/trackers/core/botsort/tracker.py:297]: done

File: src/trackers/cli/track.py (hunk 16 above)
Comment (AlexBodner) [reply src/trackers/cli/track.py:68]: yes

File: src/trackers/core/reid/appearance.py (hunk 20 above)
Comment (AlexBodner) [reply src/trackers/core/reid/appearance.py:133]: removing the skip

File: src/trackers/core/botsort/tracker.py (hunk 18 above)
Comment (AlexBodner) [reply src/trackers/core/botsort/tracker.py:222]: ok, it was 80 i think, but lets do that in a future PR so that those changes are not mixed with this big feature

File: docs/cookbooks/how-to-add-reid-to-trackers.ipynb
Comment (tstanczyk95) [reply docs/cookbooks/how-to-add-reid-to-trackers.ipynb:1]: @Borda @AlexBodner 

Normally, we use YOLOX-based detections for the numerical result reproduction in all trackers, so that we can clearly compare with the numbers in papers, while having the same detections.

For showcases, especially visual examples in Notebooks, we are free to use whichever detector. In such cases, I think RF-DETR would be suitable here to demonstrate the whole pipeline that we are proposing and providing: from detections to re-ID in tracking.

File: src/trackers/core/reid/fusion.py (hunk 6 above)
Comment (tstanczyk95) [reply src/trackers/core/reid/fusion.py:59]: @AlexBodner thanks a lot! Yes, it's better to disable it as you said. Otherwise, we might get unintended behaviour, as the IoU-based proximity was designed with standard IoU in mind and not its variants.

File: src/trackers/core/reid/appearance.py (hunk 22 above)
Comment (AlexBodner) [reply src/trackers/core/reid/appearance.py:186]: added

File: src/trackers/core/reid/appearance.py (hunk 21 above)
Comment (AlexBodner) [reply src/trackers/core/reid/appearance.py:149]: the consumer (sample_appearance_distances) needs the whole set materialised to compute pairwise
distances, so a generator would move the np.stack calls out to each caller without saving
any memory. I left it returning arrays. If consumer changes in the future we can revisit it

File: src/trackers/core/reid/thresholds.py (hunk 23 above)
Comment (AlexBodner) [reply src/trackers/core/reid/thresholds.py:399]: Done. Split the plotting into core/reid/plots.py:

File: .github/workflows/ci-reid-prerelease.yml (hunk 10 above)
Comment (AlexBodner) [reply .github/workflows/ci-reid-prerelease.yml:31]: done

File: docs/assets/reid/mot17-fastreid-appearance-distances-vs-gap.png
Comment (AlexBodner) [reply docs/assets/reid/mot17-fastreid-appearance-distances-vs-gap.png:1]: They are matplotlib figures from a calibration run over MOT17 and SoccerNet with
real encoders, so regenerating them needs the datasets and a GPU pass rather than just a
different savefig format. I would rather do it on the next calibration run than ship
re-exported approximations.

File: docs/api/reid.md (hunk 11 above)
Comment (AlexBodner) [reply docs/api/reid.md:2]: no, no other api doc has name

File: docs/guides/reid.md (hunk 7 above)
Comment (AlexBodner) [reply docs/guides/reid.md:44]: true, now pointing it out

File: docs/api/reid.md (hunk 12 above)
Comment (AlexBodner) [reply docs/api/reid.md:47]: forgot to reply to this one, we have that in guides/reid.md at PR #571

File: docs/cookbooks/how-to-add-reid-to-trackers.ipynb
Comment (AlexBodner) [reply docs/cookbooks/how-to-add-reid-to-trackers.ipynb:1]: Im resolving this comment, but we can discuss this for the future @Borda @tstanczyk95

File: docs/guides/reid.md (hunk 15 above)
Comment (AlexBodner) [reply docs/guides/reid.md]: i was running out and ran out of Colab compute units, will add them progressively
